tokio/src/net/pipe.rs

```diff
@@ -0,0 +1,898 @@
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// // Create a new FIFO file.
+    /// mkfifo(FIFO_NAME, Mode::S_IRWXU)?;
+    ///
+    /// // Wait for a reader to open the file.
+    /// let tx = loop {
+    ///     match Sender::open(FIFO_NAME) {
+    ///         Ok(tx) => break tx,
+    ///         Err(e) if e.raw_os_error() == Some(libc::ENXIO) => {},
+    ///         Err(e) => return Err(e.into()),
+    ///     }
+    ///
+    ///     time::sleep(Duration::from_millis(50)).await;
+    /// };
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open<P>(path: P) -> io::Result<Sender>
+    where
+        P: AsRef<Path>,
+    {
+        Sender::open_internal(path.as_ref(), false)
+    }
+
+    /// Creates a new `Sender` from a FIFO file without a present reader.
+    ///
+    /// This function will open the FIFO file at the specified path and associate
+    /// the pipe with the default event loop for writing.
+    ///
+    /// Unlike [`open`], this function will not fail if no one has this pipe open
+    /// for reading. This is done by opening the FIFO file with access to both
+    /// reading and writing. Note that the behavior of such operation is not defined
+    /// by the POSIX standard and is only guaranteed to work on Linux.
+    ///
+    /// [`open`]: Self::open
+    ///
+    /// # Errors
+    ///
+    /// Fails with any standard OS error if it occurs.
+    ///
+    /// # Examples
+    ///
+    /// ```no_run
+    /// use tokio::io::AsyncWriteExt;
+    /// use tokio::net::pipe::Sender;
+    /// use nix::{unistd::mkfifo, sys::stat::Mode};
+    /// # use std::error::Error;
+    ///
+    /// const FIFO_NAME: &str = "path/to/a/new/fifo";
+    ///
+    /// # async fn dox() -> Result<(), Box<dyn Error>> {
+    /// // Create a new FIFO file.
+    /// mkfifo(FIFO_NAME, Mode::S_IRWXU)?;
+    ///
+    /// // `Sender::open` would fail here, since there is no open reading end.
+    /// let mut tx = Sender::open_rw(FIFO_NAME)?;
+    /// // We can asynchronously write to the pipe before any readers.
+    /// tx.write_all(b"hello world").await?;
+    /// # Ok(())
+    /// # }
+    /// ```
+    ///
+    /// # Panics
+    ///
+    /// This function panics if it is not called from within a runtime with
+    /// IO enabled.
+    ///
+    /// The runtime is usually set implicitly when this function is called
+    /// from a future driven by a tokio runtime, otherwise runtime can be set
+    /// explicitly with [`Runtime::enter`](crate::runtime::Runtime::enter) function.
+    pub fn open_rw<P>(path: P) -> io::Result<Sender>
```

Comment:
Why not lock this behind `#[cfg(linux)]`?

Reply:
It's probably a good idea, but while only Linux explicitly allows using FIFOs in `O_RDWR` access mode, it may also be possible on other systems. For example, it seems possible on [FreeBSD](https://github.com/freebsd/freebsd-src/blob/373ffc62c158e52cde86a5b934ab4a51307f9f2e/tests/sys/fifo/fifo_io.c#L1292-L1371), but it's lacking documentation or I couldn't find it. So, an experienced user may want to use `open_wr` on other systems than Linux. 

But I think locking it to Linux is the better choice since it will probably prevent more mistakes, I will change it in the next iteration.

Reply:
We could also start with Linux and allow this on other platforms as we find out more